File: src/deploy-commands.js
import dotenv from 'dotenv';
dotenv.config();

import fs from 'fs';
import path from 'path';
import { fileURLToPath, pathToFileURL } from 'url';
import { REST } from '@discordjs/rest';
import { Routes } from 'discord-api-types/v10';

const token = process.env.DISCORD_TOKEN;
const clientId = process.env.CLIENT_ID;
const guildId = process.env.GUILD_ID; // optional: if provided, registers to a guild for instant update

if (!token || !clientId) {
  console.error('Veuilllez définir DISCORD_TOKEN et CLIENT_ID dans .env avant de déployer les commandes.');
  process.exit(1);
}

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const commandsPath = path.join(__dirname, 'slash-commands');

const commands = [];
if (fs.existsSync(commandsPath)) {
  const files = fs.readdirSync(commandsPath).filter(f => f.endsWith('.js'));
  for (const file of files) {
    try {
      const mod = await import(pathToFileURL(path.join(commandsPath, file)).href);
      const command = mod.default ?? mod;
      if (command && command.data) {
        commands.push(command.data.toJSON());
      } else {
        console.warn(`Le fichier ${file} n'exporte pas de 'data' (SlashCommandBuilder).`);
      }
    } catch (err) {
      console.error('Erreur en important', file, err.message || err);
    }
  }
}

const rest = new REST({ version: '10' }).setToken(token);

try {
  if (guildId) {
    console.log(`Déploiement des ${commands.length} commandes au guild ${guildId}...`);
    const data = await rest.put(Routes.applicationGuildCommands(clientId, guildId), { body: commands });
    console.log('Commandes déployées (guild):', data.length);
  } else {
    console.log(`Déploiement des ${commands.length} commandes en global (cela peut prendre jusqu'à 1 heure)...`);
    const data = await rest.put(Routes.applicationCommands(clientId), { body: commands });
    console.log('Commandes déployées (global):', data.length);
  }
} catch (err) {
  console.error('Erreur lors du déploiement des commandes:', err);
}

File: Python/deploy_commands.py
"""Script de déploiement des slash-commands (version Python).

Il parcourt le dossier `Python/slash_commands` et construit des définitions
simples (name + description). Ensuite il appelle l'API REST Discord pour
déployer en guild ou global selon les variables d'environnement.

Ne nécessite pas discord.py pour fonctionner (utilise urllib pour REST).
"""
from dotenv import load_dotenv
load_dotenv()

import os
import sys
import json
import urllib.request
import urllib.error
from pathlib import Path
import importlib.util

from logger import Logger

LOG = Logger()

TOKEN = os.getenv('DISCORD_TOKEN')
CLIENT_ID = os.getenv('CLIENT_ID')
GUILD_ID = os.getenv('GUILD_ID')
DEPLOY_ALL = os.getenv('DEPLOY_ALL_GUILDS', '').lower() == 'true'

BASE_DIR = Path(__file__).resolve().parent
SLASH_DIR = BASE_DIR / 'slash_commands'

if not TOKEN or not CLIENT_ID:
    LOG.error('DISCORD_TOKEN et CLIENT_ID sont requis pour déployer les commandes.')
    sys.exit(1)


def _http_request(method: str, url: str, data=None):
    headers = {
        'Authorization': f'Bot {TOKEN}',
        'Content-Type': 'application/json'
    }
    body = None
    if data is not None:
        body = json.dumps(data).encode('utf8')
    req = urllib.request.Request(url, data=body, headers=headers, method=method)
    try:
        with urllib.request.urlopen(req, timeout=30) as resp:
            raw = resp.read()
            try:
                return json.loads(raw.decode('utf8'))
            except Exception:
                return raw.decode('utf8')
    except urllib.error.HTTPError as e:
        try:
            body = e.read().decode('utf8')
            LOG.error(f'HTTP {e.code} {e.reason}: {body}')
        except Exception:
            LOG.error(f'HTTP {e.code} {e.reason}')
        raise
    except Exception as e:
        LOG.error(f'Erreur HTTP: {e}')
        raise


def collect_commands():
    commands = []
    if not SLASH_DIR.exists():
        LOG.warn('Aucun dossier `Python/slash_commands` — rien à déployer.')
        return commands
    for p in SLASH_DIR.iterdir():
        if not p.is_file() or p.suffix != '.py' or p.stem.startswith('_'):
            continue
        try:
            spec = importlib.util.spec_from_file_location(f'slash_commands.{p.stem}', str(p))
            mod = importlib.util.module_from_spec(spec)
            spec.loader.exec_module(mod)
            name = getattr(mod, 'name', None)
            desc = getattr(mod, 'description', '')
            data = getattr(mod, 'data', None)
            if data and isinstance(data, dict):
                commands.append(data)
            elif name:
                commands.append({
                    'name': name,
                    'type': 1,  # CHAT_INPUT
                    'description': desc or 'Commande',
                    'options': []
                })
            else:
                LOG.warn(f'Fichier {p.name} ignoré: pas de `name` exporté')
        except Exception as e:
            LOG.error(f'Erreur en important {p.name}: {e}')
    return commands


def deploy_to_guilds(commands):
    rest_base = 'https://discord.com/api/v10'
    if GUILD_ID and GUILD_ID != 'ALL' and not DEPLOY_ALL:
        url = f"{rest_base}/applications/{CLIENT_ID}/guilds/{GUILD_ID}/commands"
        LOG.info(f'Déploiement de {len(commands)} commandes au guild {GUILD_ID}...')
        data = _http_request('PUT', url, commands)
        LOG.info(f'Commandes déployées (guild): {len(data) if isinstance(data, list) else 1}')
        return

    if GUILD_ID == 'ALL' or DEPLOY_ALL:
        LOG.info('Récupération des guilds du bot pour déploiement sur chacune...')
        guilds = _http_request('GET', f'https://discord.com/api/v10/users/@me/guilds')
        if not isinstance(guilds, list) or not guilds:
            LOG.warn('Aucune guild trouvée pour le bot.')
            return
        for g in guilds:
            gid = g.get('id')
            try:
                LOG.info(f'Déploiement pour guild {gid}...')
                url = f"{rest_base}/applications/{CLIENT_ID}/guilds/{gid}/commands"
                data = _http_request('PUT', url, commands)
                LOG.info(f'  OK — {len(data) if isinstance(data, list) else 1} commandes déployées')
            except Exception as e:
                LOG.error(f'Erreur pour guild {gid}: {e}')
        return

    # global deploy
    url = f'https://discord.com/api/v10/applications/{CLIENT_ID}/commands'
    LOG.info(f'Déploiement global de {len(commands)} commandes (peut prendre ~1h)...')
    data = _http_request('PUT', url, commands)
    LOG.info(f'Commandes déployées (global): {len(data) if isinstance(data, list) else 1}')


def main():
    cmds = collect_commands()
    LOG.info(f'Prêt à déployer {len(cmds)} slash-commands')
    if not cmds:
        LOG.warn('Aucune commande collectée — rien à faire.')
        return
    deploy_to_guilds(cmds)


if __name__ == '__main__':
    main()
